### ISSUE TITLE
EFCORE 5.0: ZERO 'writeOperations' created when inserting NULL in Dependent Entity columns.

### ISSUE THREAD

**Issue by rkaushik15** [dotnet/efcore#23304]
### Description

I am part of a team developing an EFCORE provider for Oracle DB compatible with EFCORE 5.0 and running into this issue.
Column names are needed to generate the INSERT statement in Oracle DB but we are not getting that information from the Relational Layer when trying to insert 'null' in dependent entity columns as follows:
 
```
      MyContext22054 db = new MyContext22054();
      db.AddRange(
        new User22054 { Contact = null, Data = null }
      );
      db.SaveChanges();
```

We do not get any 'writeOperations' generated, and thus we have no column information which we need to create the INSERT SQL.

SQL Server is able to avoid this issue as it has support to specify 'DEFAULT VALUES' when inserting, 
The statement can be something like this:

```
INSERT INTO [User22054]
DEFAULT VALUES;
``` 
This 'DEFAULT VALUES' is used when there are 0 'writeOperations' (refer to **AppendValuesHeader()** methd in the **UpdateSqlGenerator()** class).

However, for Oracle DB, we require the Column Name, as there is no support for specifying 'DEFAULT VALUES'.
Thus to create the INSERT statement properly, we need the 'writeOperations'  so that we can get the column names and insert 'null' in those.

In case we try to insert 'null' in non-dependent entity columns, the write operation is created, and a SQL that inserts 'null' can be generated. The behavior should be the same when 'null' is inserted in the dependent entity columns, i.e., **'writeOperations' should be generated, specifying the column names that we are adding with 'null' values.**

NOTE:  I'm referring to the **'writeOperations'** in the **AppendInsertCommand()** method in the **UpdateSqlGenerator()** class. This is the relevant method responsible to generate the insert command according to my analysis.

### Code
Please refer to the attached SQL Server test case. To see the 'writeOperations', you can debug **AppendInsertCommand()** method in the **UpdateSqlGenerator()** class.
[testcase.zip](https://github.com/dotnet/efcore/files/5536192/testcase.zip)

### Provider and version information

EF Core version: 5.0
Target framework: netcoreapp3.1
Operating system: Windows 10
IDE: Visual Studio, Version 16.8.0 Preview 4.0

**Comment by roji** [comment]
@rkaushik15 apologies for taking so long to answer this - we've been quite busy with the recent 5.0 release.

I understand that you're looking to customize the INSERT statements generated by EF Core, so that instead of `DEFAULT VALUES`, you generate null values for Oracle database-generated columns.

The writeOperations are empty on the ModificationCommand because when the ColumnModifications [are generated](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Update/ModificationCommand.cs#L259), any store-generated column [is assumed to be a read-only operation](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Update/ModificationCommand.cs#L316) (not requiring a write) - that's usually how it works in other database. You could override GenerateColumnModifications (by having your own CommandBatchPreparer, which constructs your own ModificationCommand) to set these ColumnModifications as writable. Unfortunately that would force you to duplicate quite a bit of code from EF.

An easier option may be to hook in later, overriding UpdateSqlGenerator.AppendInsertOperation and giving the base implementation a rewritten ModificationCommand, where the read-only ColumnModifications for store-generated columns have been replaced with write operations. You should be able to identify ColumnModifications for store-generated columns by looking at the Entry's IsStoreGenerated value.

Otherwise, @AndriySvyryd probably has more insights to add.

BTW out of curiosity, if your plan is to send a null value as an indication that the column should get generated by the database, what happens with a store-generated column whose default is non-null, and where the user wants to insert null? In other words, how would the database distinguish between null as the actual value that the user wants to insert, and null as an indication that the database should generate the value? This is why in other database a special construct is used for the latter.

**Comment by rkaushik15** [comment]
Hi @roji 
Thanks for your response and congratulations on the successful launch of EFCORE 5.0 :)

>I understand that you're looking to customize the INSERT statements generated by EF Core, so that instead of DEFAULT VALUES, you generate null values for Oracle database-generated columns.

A point of clarification here, in this case we are NOT trying to insert 'null' in the store generated column. We are inserting 'null' in some other columns of this entity which has some (other) store generated columns.

A simplified explanation here, 'User' has an _IDENTITY_ Column 'Id', and owns an entity 'Contact'.

```
private class User
    {
        public int Id { get; set; }
        public Contact contact { get; set; }
     }
```

We are trying to insert 'null' Contact, and not the store generated column 'Id' 

```
db.AddRange(
        new User { Contact = null }
      );
      db.SaveChanges();
```

>The writeOperations are empty on the ModificationCommand because when the ColumnModifications are generated, any store-generated column is assumed to be a read-only operation (not requiring a write) - that's usually how it works in other database

No writeOperation and only readOperation generated for 'Id' works just fine (as explained by you above). 
However, **shouldn't there be writeOperations corresponding to the 'Contact' related columns that have 'null' specified as the value to be inserted**? 

Looking at another example, suppose 'User' has an _IDENTITY_ "Id" and another column which is **string** and not a dependent entity.

```
 private class User
    {
      public int Id { get; set; }
      public string COL1 { get; set; }
    }
```

If we insert 'null' in 'COL1' as follows:

```
db.AddRange(
        new User { COL1= null }
      );
      db.SaveChanges();
```

Now we get a readOperation for 'Id' **AND** a writeOperation, that says insert 'null' in 'COL1', which seems to be the correct behaviour. 

The issue is not with inserting 'null' in store generated columns and hence the solution proposed by you above would not be valid I suppose.

**Comment by ajcvickers** [comment]
Ping @roji

**Comment by roji** [comment]
@rkaushik15 thanks for the added details.

If I understand the above correctly, you're asking why a NULL gets sent for an unset string property, but not for an unset owned entity... I can indeed see a difference in behavior there. However, if anything, it seems to me that EF shouldn't send a NULL for the unset string property, rather than starting to send nulls for all the properties of the unset owned entity.

But I think the fundamental issue here is different... What would happen in your case, if you have some string column (no owned entity), just like your COL1 above, but the property is configured with a default value? In that case, when the property is unset, EF definitely cannot send a NULL (since we want the column to get the default value, which isn't NULL).

In other words, is there really no way to insert a new row in Oracle without specifying any columns (because we want all columns to have their default values)?

**Comment by roji** [comment]
@rkaushik15 any feedback on this? I'm curious to understand the Oracle situation better.

**Comment by rkaushik15** [comment]
@roji sorry for the delay in response.

>What would happen in your case, if you have some string column (no owned entity), just like your COL1 above, but the property is configured with a default value?

Oracle has two modes to deal with default values. We can either specify the default columns using 'DEFAULT' or 'DEFAULT ON NULL' when creating the columns. Thus, depending on how the column is configured, when EFCORE sends 'null' to be inserted in a column with a default value, it will either insert 'null' or the default value. 

>In other words, is there really no way to insert a new row in Oracle without specifying any columns (because we want all columns to have their default values)?

No, there's no way to insert a row without specifying columns. For an insert statement in Oracle, we have to specify a column. If we wish to insert the defaults in all columns we specify all the columns and put 'default' in the VALUES field.

**Comment by roji** [comment]
> Oracle has two modes to deal with default values. We can either specify the default columns using 'DEFAULT' or 'DEFAULT ON NULL' when creating the columns.

So if I understand correctly, columns with 'DEFAULT ON NULL' basically can never contain null, since attempting to insert NULL into them inserts the configured default value instead?

Without knowing much more about the Oracle context, would it make sense to simply support the first type, and to modify the update pipeline to send `default`in the VALUES field? Or is there some reason why `DEFAULT ON NULL` must also be supported?

In any case, both of the above seem like they'd require some customization to ModificationCommand and/or ColumnModification. @AndriySvyryd do you have any input here?

**Comment by rkaushik15** [comment]
>So if I understand correctly, columns with 'DEFAULT ON NULL' basically can never contain null, since attempting to insert NULL into them inserts the configured default value instead?

That is correct!

>Without knowing much more about the Oracle context, would it make sense to simply support the first type, and to modify the update pipeline to send default in the VALUES field? 

It does make sense. In fact, that is what we currently do in our implementation. We check the writeOperations and if it is a default column and the value to be inserted is null, we can send 'default'. The problem arises when the unset property is an owned entity. In that case, we have no writeOperations and thus we do not have the column name that we need to create the SQL query (and as I mentioned before, the Oracle insert query needs to have column name). We do not wish to support the insertion of 'null' per se, but need the column info of the unset property be it owned entity or not.

As you've noted the behavior is different for when the unset property is an owned entity vs. when it is (for example) a string. Do you see some merit in making this behavior consistent for both cases?

**Comment by roji** [comment]
> As you've noted the behavior is different for when the unset property is an owned entity vs. when it is (for example) a string. Do you see some merit in making this behavior consistent for both cases?

It's an interesting question, and it's worth keeping in mind that there's a performance aspect here... While not sending NULLs may seem better when they're not necessary, it would also mean different SQL statements for different data combinations, which  may have an impact on plan caches at the server. May initial untested assumption is that it's probably better to keep sending nulls for unset properties so as to have identical SQL statements. The same logic could be used to argue to send nulls for owned entity properties, though I think that's considerably weaker.

Aside from that, since sending a null and not sending anything at all is functionally the same (except for Oracle), I'm not sure there's much value in making things consistent here... I'll raise this with the team though.

> The problem arises [...]

I haven't checked whether this is fully feasible, but it seems like it should be possible to find out the columns in question by checking the Entries on the ModificationCommand, and to get to the EntityTypes from there. You could then find relevant properties (auto-generated, not being explicitly written) and add them to the ColumnModification list. This could be a bit elaborate, but it may be worth checking it out.

**Comment by roji** [comment]
@rkaushik15 good news, I've consulted with team, and the behavior whereby unset properties of owned entities aren't sent turns out to be a bug. I've filed #23824 to track this - the fix should make it easier for you to intercept all properties as you wanted.

Am going to close this as I think that's the right outcome for the Oracle provider, but don't hesitate to post back here for any reason.

**Comment by roji** [comment]
Duplicate of #23824

**Comment by rkaushik15** [comment]
Hi @roji 
Great! Thanks for the update :)
